Sources/SwiftParser/Lexer/Cursor.swift

```diff
@@ -1919,7 +1957,37 @@ extension Lexer.Cursor {
     return Lexer.Result(.backtick)
   }
 
-  mutating func lexOperatorIdentifier(sourceBufferStart: Lexer.Cursor) -> Lexer.Result {
+  /// Attempt to lex a postfix '!' or '?'.
+  mutating func lexPostfixOptionalChain(sourceBufferStart: Lexer.Cursor) -> Lexer.Result? {
+    // Must be left bound, otherwise this isn't postfix.
+    guard self.isLeftBound(sourceBufferStart: sourceBufferStart) else { return nil }
+
+    var transition: Lexer.StateTransition?
+    if previousKeyword == .try {
+      // If we have 'try' as the previous keyword kind, we have `try?` or `try!`
+      // and need to transition into the state where we prefer lexing a regex
+      // literal over a binary operator. See the comment in
+      // `tryScanOperatorAsRegexLiteral` for more info.
+      transition = .push(newState: .preferRegexOverBinaryOperator)
+    }
+    let kind: RawTokenKind = {
+      switch self.peek() {
+      case UInt8(ascii: "!"):
+        return .exclamationMark
+      case UInt8(ascii: "?"):
+        return .postfixQuestionMark
+      default:
+        fatalError("Must be at '!' or '?'")
```

Comment:
Personally `fatalError` feels more "this shouldn't be reachable" whereas `preconditionFailure` feels more "you violated the contract of an API". This definitely fits into the former, but I'm happy to change if you want.

Reply:
I just think we should use one or the other because otherwise I just always start wondering: Why did someone choose `fatalError` here instead of `preconditionFailure`, which is most of the time completely pointless to wonder about.